In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [4]:
dataset = pd.read_csv("Restaurant_Reviews.tsv",delimiter='\t',quoting=3)

In [6]:
dataset

Review  Liked
0                             Wow... Loved this place.      1
1                                   Crust is not good.      0
2            Not tasty and the texture was just nasty.      0
3    Stopped by during the late May bank holiday of...      1
4    The selection on the menu was great and so wer...      1
5       Now I am getting angry and I want my damn pho.      0
6                Honeslty it didn't taste THAT fresh.)      0
7    The potatoes were like rubber and you could te...      0
8                            The fries were great too.      1
9                                       A great touch.      1
10                            Service was very prompt.      1
11                                  Would not go back.      0
12   The cashier had no care what so ever on what I...      0
13   I tried the Cape Cod ravoli, chicken, with cra...      1
14   I was disgusted because I was pretty sure that...      0
15   I was shocked because no signs indicate cash o...      0
16                                 Highly recommended.      1
17              Waitress was a little slow in service.      0
18   This place is not worth your time, let alone V...      0
19                                did not like at all.      0
20                                 The Burrittos Blah!      0
21                                  The food, amazing.      1
22                               Service is also cute.      1
23   I could care less... The interior is just beau...      1
24                                  So they performed.      1
25   That's right....the red velvet cake.....ohhh t...      1
26          - They never brought a salad we asked for.      0
27   This hole in the wall has great Mexican street...      1
28   Took an hour to get our food only 4 tables in ...      0
29                   The worst was the salmon sashimi.      0
..                                                 ...    ...
970  I immediately said I wanted to talk to the man...      0
971                    The ambiance isn't much better.      0
972  Unfortunately, it only set us up for disapppoi...      0
973                              The food wasn't good.      0
974  Your servers suck, wait, correction, our serve...      0
975      What happened next was pretty....off putting.      0
976  too bad cause I know it's family owned, I real...      0
977               Overpriced for what you are getting.      0
978               I vomited in the bathroom mid lunch.      0
979  I kept looking at the time and it had soon bec...      0
980  I have been to very few places to eat that und...      0
981  We started with the tuna sashimi which was bro...      0
982                            Food was below average.      0
983  It sure does beat the nachos at the movies but...      0
984       All in all, Ha Long Bay was a bit of a flop.      0
985  The problem I have is that they charge $11.99 ...      0
986  Shrimp- When I unwrapped it (I live only 1/2 a...      0
987     It lacked flavor, seemed undercooked, and dry.      0
988  It really is impressive that the place hasn't ...      0
989  I would avoid this place if you are staying in...      0
990  The refried beans that came with my meal were ...      0
991         Spend your money and time some place else.      0
992  A lady at the table next to us found a live gr...      0
993            the presentation of the food was awful.      0
994           I can't tell you how disappointed I was.      0
995  I think food should have flavor and texture an...      0
996                           Appetite instantly gone.      0
997  Overall I was not impressed and would not go b...      0
998  The whole experience was underwhelming, and I ...      0
999  Then, as if I hadn't wasted enough of my life ...      0

[1000 rows x 2 columns]

using a tsv file(and not a csv) file because reviews can themeselves contain commas but never tabs

ignoring quotes as they may add an extra layer of complexity

Goal:- create a model that predicts wether a  new review is positive or negative

### Cleaning the text:-
preparing it for use in the model

cleaning the first review for an example and will later be iterating through all the reviews

In [12]:
import re
# the regular expression operations module of python

In [14]:
dataset['Review'][0]

'Wow... Loved this place.'

In [15]:
review = re.sub('[^a-zA-Z]',' ',dataset['Review'][0])

read the docstring for the re.sub function

In [16]:
review

'Wow    Loved this place '

In [18]:
review = review.lower()

In [19]:
review

'wow    loved this place '

will be using a sparse matrix hence will remove unnecesary words - "this" in this case

In [20]:
import nltk

In [21]:
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to C:\Users\Raj
[nltk_data]     Patil\AppData\Roaming\nltk_data...
[nltk_data]   Unzipping corpora\stopwords.zip.


True

In [26]:
from nltk.corpus import stopwords

splitting the review in different words so as to make it an iterable list

In [23]:
review = review.split()

In [24]:
review

['wow', 'loved', 'this', 'place']

In [31]:
review = [word for word in review if not word in set(stopwords.words('english'))]

In [32]:
review

['wow', 'loved', 'place']

using set of stopwords rather than a list because python goes through sets faster than lists

### Stemming the review

In essence loved will be recorded as love because different versions derived from the same root will give the same idea to the model hence it helps in making the sparsity matrix more efficiently

In [36]:
from nltk.stem.porter import PorterStemmer

In [37]:
ps = PorterStemmer()

will do this before removing the unnecessary words

hence resetting the review, and implementing the expression cleaning and stemming in the desired order

In [47]:
review = re.sub('[^a-zA-z]',' ',dataset['Review'][0])
review = review.lower()
review = review.split()
review = [ps.stem(word) for word in review if not word in set(stopwords.words('english'))]

In [48]:
review

['wow', 'love', 'place']

Now, We're good

now remerging the list into one string

In [49]:
review = " ".join(review)

In [50]:
review

'wow love place'

now doing this for all the reviews iterating through the dataset

In [54]:
corpus = []#stores the modified reviews
for i in range(0,1000):
    review = re.sub('[^a-zA-Z]',' ',
                    dataset['Review'][i])
    review = review.lower()
    review = review.split()
    review = [ps.stem(word) for word in review if not
              word in set(stopwords.words('english'))]
    review = ' '.join(review)
    corpus.append(review)

    

In [58]:
pd.DataFrame(corpus)

0
0                                       wow love place
1                                           crust good
2                                   tasti textur nasti
3    stop late may bank holiday rick steve recommen...
4                              select menu great price
5                              get angri want damn pho
6                                  honeslti tast fresh
7    potato like rubber could tell made ahead time ...
8                                            fri great
9                                          great touch
10                                       servic prompt
11                                       would go back
12        cashier care ever say still end wayyy overpr
13          tri cape cod ravoli chicken cranberri mmmm
14                      disgust pretti sure human hair
15                               shock sign indic cash
16                                    highli recommend
17                          waitress littl slow servic
18                      place worth time let alon vega
19                                                like
20                                       burritto blah
21                                           food amaz
22                                    servic also cute
23                     could care less interior beauti
24                                             perform
25               right red velvet cake ohhh stuff good
26                             never brought salad ask
27   hole wall great mexican street taco friendli s...
28   took hour get food tabl restaur food luke warm...
29                                worst salmon sashimi
..                                                 ...
970  immedi said want talk manag want talk guy shot...
971                                ambianc much better
972                  unfortun set us disapppoint entre
973                                          food good
974        server suck wait correct server heimer suck
975                             happen next pretti put
976    bad caus know famili own realli want like place
977                                         overpr get
978                           vomit bathroom mid lunch
979     kept look time soon becom minut yet still food
980      place eat circumst would ever return top list
981    start tuna sashimi brownish color obvious fresh
982                                        food averag
983  sure beat nacho movi would expect littl bit co...
984                               ha long bay bit flop
985  problem charg sandwich bigger subway sub offer...
986    shrimp unwrap live mile brushfir liter ice cold
987                     lack flavor seem undercook dri
988                         realli impress place close
989                       would avoid place stay mirag
990         refri bean came meal dri crusti food bland
991                         spend money time place els
992  ladi tabl next us found live green caterpillar...
993                                    present food aw
994                                    tell disappoint
995                      think food flavor textur lack
996                             appetit instantli gone
997                       overal impress would go back
998  whole experi underwhelm think go ninja sushi n...
999  wast enough life pour salt wound draw time too...

[1000 rows x 1 columns]

### The bag of words model 

<i> a matrix with lot of zeros is a sparse matrix</i>

creating a sparse matrix using tokenisation

removing the rare variables in the corpus eg:- names for reducing the sparsity of the matrix

IN ESSENCE<br>
this is a classification problem, we get all these cleaned and stemmed words as our categorical variables and we are predicting wether the output is a 1 or a 0


### creating the model now

In [62]:
from sklearn.feature_extraction.text import CountVectorizer

In [63]:
cv = CountVectorizer()

 note that we are not using any parameters now, but have a look at them,  most of the manual cleaning of the text can be done via triggering some of these parameters
<br><br>
But, manual cleaning is preferable as it gives us more control over what we want to clean/include


### creating the sparse matrix

In [66]:
X = cv.fit_transform(corpus).toarray()

In [68]:
X.shape

(1000, 1565)

so we have reduced it down to a total of  1565 separate words(tokens) 

if the number of words are a lot, we can play with some of the parameters of the CountVectorizer - see for yourself

In [71]:
cv = CountVectorizer(max_features=1500)
X = cv.fit_transform(corpus)

In [79]:
X.shape

(1000, 1500)

In [80]:
X = X.toarray()

this is a classification problem now

In [82]:
Y = dataset.iloc[:,1].values

In [83]:
from sklearn.model_selection import train_test_split

In [84]:
X_train,X_test,Y_train,Y_test = train_test_split(X,Y,test_size=0.20,random_state=0)


### the most common models used for NLP classification are:-  
<ul>
    <li>Naïve Bayes</li>
    <li>Decision Trees</li>
    <li>Random Forests</li>
    </ul>

I'm using the Naïve Bayes here

In [85]:
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train,Y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

In [87]:
Y_pred = classifier.predict(X_test)

In [88]:
from sklearn.metrics import confusion_matrix
cm = confusion_matrix(Y_test,Y_pred)

In [91]:
cm

array([[55, 42],
       [12, 91]], dtype=int64)

In [94]:
accuracy = (55+91)/200
accuracy*100

73.0

Not bad for this small of a test set,
so, will improve this later using some other model and parameter tuning